File: opt/claude-code/hooks/langfuse-trace.ts
#!/usr/bin/env -S -- node

import type { BaseHookInput } from "@anthropic-ai/claude-agent-sdk"
import type {
  ContentBlock,
  Message,
  ToolResultBlockParam,
  ToolUseBlock,
} from "@anthropic-ai/sdk/resources/messages/messages.js"
import { LangfuseSpanProcessor } from "@langfuse/otel"
import {
  propagateAttributes,
  setLangfuseTracerProvider,
  startActiveObservation,
  startObservation,
} from "@langfuse/tracing"
import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node"
import { createHash } from "node:crypto"
import { createReadStream } from "node:fs"
import { mkdir, readFile, rename, writeFile } from "node:fs/promises"
import { dirname, resolve } from "node:path"
import { env, exit, stdin } from "node:process"
import { text } from "node:stream/consumers"
import { fileURLToPath } from "node:url"

// ── Constants ────────────────────────────────────────

const MAX_CHARS = parseInt(env["CC_LANGFUSE_MAX_CHARS"] ?? "20000", 10)
const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "../../..")
const SESSIONS_DIR = resolve(ROOT, "var/sessions")

// ── Types ────────────────────────────────────────────

type Config = {
  publicKey: string
  secretKey: string
  host: string
}

type ReaderState = {
  offset: number
  buffer: string
  turnCount: number
}

type TranscriptLine = {
  type?: "user" | "assistant"
  message?: Partial<Message> & { role?: string }
  content?: Message["content"] | string
}

type TruncMeta = {
  truncated: boolean
  orig_len: number
  kept_len?: number
  sha256?: string
}

type ToolCall = {
  id: string
  name: string
  input: unknown
  output?: string | null
  output_meta?: TruncMeta
}

type Turn = {
  userMsg: TranscriptLine
  assistantMsgs: TranscriptLine[]
  toolResults: Map<string, unknown>
}

// ── Disposable helpers ───────────────────────────────

const disposable = <T extends { end(): void }>(obs: T) =>
  Object.assign(obs, { [Symbol.dispose]: () => obs.end() })

const timed = (label: string) => {
  const start = performance.now()
  log({ level: "debug", msg: `${label} started` })
  return {
    [Symbol.dispose]() {
      const dur = ((performance.now() - start) / 1000).toFixed(2)
      log({ level: "info", msg: `${label} completed in ${dur}s` })
    },
  }
}

// ── Logging ──────────────────────────────────────────

const isDebug = env["CC_LANGFUSE_DEBUG"]?.toLowerCase() === "true"

const log = ({
  level,
  msg,
}: {
  level: "debug" | "info" | "error"
  msg: string
}) => {
  if (level === "debug" && !isDebug) return
  console.error(`[${level}] ${msg}`)
}

// ── Input ────────────────────────────────────────────

const readConfig = (): Config | null => {
  if (env["TRACE_TO_LANGFUSE"]?.toLowerCase() !== "true") return null

  const publicKey = env["CC_LANGFUSE_PUBLIC_KEY"] ?? env["LANGFUSE_PUBLIC_KEY"]
  const secretKey = env["CC_LANGFUSE_SECRET_KEY"] ?? env["LANGFUSE_SECRET_KEY"]
  if (!publicKey || !secretKey) return null

  const host =
    env["CC_LANGFUSE_BASE_URL"] ??
    env["LANGFUSE_BASE_URL"] ??
    "https://cloud.langfuse.com"

  return { publicKey, secretKey, host }
}

const readPayload = async () => {
  const data = await text(stdin)
  if (!data.trim()) return null
  return JSON.parse(data) as BaseHookInput & { hook_event_name?: string }
}

// ── State ────────────────────────────────────────────

const openState = async (sessionId: string) => {
  const path = resolve(SESSIONS_DIR, `${sessionId}.langfuse.json`)

  const state = await readFile(path, "utf-8")
    .then((data) => {
      const raw = JSON.parse(data)
      return {
        offset: Number(raw["offset"] ?? 0),
        buffer: String(raw["buffer"] ?? ""),
        turnCount: Number(raw["turn_count"] ?? 0),
      }
    })
    .catch(() => ({ offset: 0, buffer: "", turnCount: 0 }))

  return Object.assign(state, {
    async [Symbol.asyncDispose]() {
      const tmp = `${path}.tmp`
      const json = JSON.stringify(
        {
          offset: state.offset,
          buffer: state.buffer,
          turn_count: state.turnCount,
          updated: new Date().toISOString(),
        },
        null,
        2,
      )

      await mkdir(dirname(path), { recursive: true })
        .then(() => writeFile(tmp, json, "utf-8"))
        .then(() => rename(tmp, path))
        .catch((e) => log({ level: "error", msg: String(e) }))
    },
  })
}

// ── Message vocabulary ───────────────────────────────

const getContent = (msg: TranscriptLine) => msg.message?.content ?? msg.content

const getRole = (msg: TranscriptLine) => {
  const role = msg.type ?? msg.message?.role
  return role === "user" || role === "assistant" ? role : undefined
}

const contentBlocks = <T extends ContentBlock | ToolResultBlockParam>(
  content: (ContentBlock | ToolResultBlockParam)[] | string | undefined,
  blockType: string,
): T[] => {
  if (!Array.isArray(content)) return []
  return content.filter(
    (x): x is T => typeof x === "object" && x !== null && x.type === blockType,
  )
}

const isToolResult = (msg: TranscriptLine) =>
  getRole(msg) === "user" &&
  contentBlocks<ToolResultBlockParam>(getContent(msg), "tool_result").length > 0

const getModel = (msg: TranscriptLine) => msg.message?.model ?? "claude"

const getMessageId = (msg: TranscriptLine) => msg.message?.id ?? undefined

const extractText = (
  content: (ContentBlock | ToolResultBlockParam)[] | string | undefined,
): string => {
  if (typeof content === "string") return content
  if (!Array.isArray(content)) return ""

  return content
    .flatMap((block) => {
      if (typeof block === "string" && block) return [block]
      if (
        typeof block === "object" &&
        block !== null &&
        "type" in block &&
        block.type === "text" &&
        "text" in block &&
        typeof block.text === "string"
      )
        return [block.text]
      return []
    })
    .join("\n")
}

const truncate = (s: string, maxChars = MAX_CHARS): [string, TruncMeta] => {
  const origLen = s.length
  if (origLen <= maxChars) return [s, { truncated: false, orig_len: origLen }]
  const head = s.slice(0, maxChars)
  return [
    head,
    {
      truncated: true,
      orig_len: origLen,
      kept_len: head.length,
      sha256: createHash("sha256").update(s, "utf-8").digest("hex"),
    },
  ]
}

// ── Parse ────────────────────────────────────────────

const readNewMessages = async (
  path: string,
  state: ReaderState,
): Promise<[TranscriptLine[], ReaderState]> => {
  const chunk = await text(
    createReadStream(path, { start: state.offset, encoding: "utf-8" }),
  ).catch((e) => {
    log({ level: "error", msg: String(e) })
    return ""
  })
  if (!chunk) return [[], state]

  const combined = state.buffer + chunk
  const lines = combined.split("\n")

  const msgs = lines.slice(0, -1).flatMap((raw) => {
    const trimmed = raw.trim()
    if (!trimmed) return []
    try {
      return [JSON.parse(trimmed) as TranscriptLine]
    } catch {
      return []
    }
  })

  return [
    msgs,
    {
      ...state,
      offset: state.offset + Buffer.byteLength(chunk),
      buffer: lines.at(-1) ?? "",
    },
  ]
}

// ── Assemble ─────────────────────────────────────────

const assembleTurns = function* (
  messages: TranscriptLine[],
): IteratorObject<Turn> {
  let userMsg: TranscriptLine | null = null
  let assistants = new Map<string, TranscriptLine>()
  let toolResults = new Map<string, unknown>()

  for (const msg of messages) {
    if (isToolResult(msg)) {
      for (const tr of contentBlocks<ToolResultBlockParam>(
        getContent(msg),
        "tool_result",
      )) {
        if (tr.tool_use_id) toolResults.set(tr.tool_use_id, tr.content)
      }
      continue
    }

    const role = getRole(msg)

    if (role === "user") {
      if (userMsg !== null && assistants.size > 0) {
        yield {
          userMsg,
          assistantMsgs: assistants.values().toArray(),
          toolResults: new Map(toolResults),
        }
      }
      userMsg = msg
      assistants = new Map()
      toolResults = new Map()
      continue
    }

    if (role === "assistant" && userMsg !== null) {
      const mid = getMessageId(msg) ?? `noid:${assistants.size}`
      assistants.set(mid, msg)
      continue
    }
  }

  if (userMsg !== null && assistants.size > 0) {
    yield {
      userMsg,
      assistantMsgs: assistants.values().toArray(),
      toolResults: new Map(toolResults),
    }
  }

  return
}

// ── Emit ─────────────────────────────────────────────

const toolCalls = (assistantMsgs: TranscriptLine[]): ToolCall[] =>
  assistantMsgs.flatMap((am) =>
    contentBlocks<ToolUseBlock>(getContent(am), "tool_use").map((tu) => ({
      id: tu.id,
      name: tu.name,
      input: tu.input,
    })),
  )

const emitTurn = ({
  sessionId,
  turnNum,
  turn,
  transcriptPath,
}: {
  sessionId: string
  turnNum: number
  turn: Turn
  transcriptPath: string
}) => {
  const [userText, userMeta] = truncate(extractText(getContent(turn.userMsg)))
  const lastAssistant = turn.assistantMsgs.at(-1) ?? turn.userMsg
  const [assistantText, assistantMeta] = truncate(
    extractText(getContent(lastAssistant)),
  )
  const model = getModel(turn.assistantMsgs[0] ?? lastAssistant)
  const calls = toolCalls(turn.assistantMsgs)

  for (const c of calls) {
    const raw = turn.toolResults.get(c.id)
    if (raw !== undefined) {
      const outStr = typeof raw === "string" ? raw : JSON.stringify(raw)
      const [outTrunc, outMeta] = truncate(outStr)
      c.output = outTrunc
      c.output_meta = outMeta
    } else {
      c.output = null
    }
  }

  const traceName = `Claude Code - Turn ${turnNum}`

  propagateAttributes({ sessionId, traceName, tags: ["claude-code"] }, () =>
    startActiveObservation(traceName, (trace) => {
      trace.update({
        input: { role: "user", content: userText },
        metadata: {
          source: "claude-code",
          session_id: sessionId,
          turn_number: turnNum,
          transcript_path: transcriptPath,
          user_text: userMeta,
        },
      })

      {
        using _ = disposable(
          startObservation(
            "Claude Response",
            {
              input: { role: "user", content: userText },
              output: { role: "assistant", content: assistantText },
              model,
              metadata: {
                assistant_text: assistantMeta,
                tool_count: calls.length,
              },
            },
            { asType: "generation" },
          ),
        )
      }

      for (const c of calls) {
        const [inObj, inMeta] =
          typeof c.input === "string" ? truncate(c.input) : [c.input, null]

        using toolObs = disposable(
          startObservation(
            `Tool: ${c.name}`,
            {
              input: inObj,
              metadata: {
                tool_name: c.name,
                tool_id: c.id,
                input_meta: inMeta,
                output_meta: c.output_meta,
              },
            },
            { asType: "tool" },
          ),
        )
        toolObs.update({ output: c.output })
      }

      for (const c of calls) {
        if (c.name === "ExitPlanMode" && c.input) {
          const planStr =
            typeof c.input === "string" ? c.input : JSON.stringify(c.input)
          const [planTrunc, planMeta] = truncate(planStr)
          using _ = disposable(
            startObservation("Plan", {
              output: planTrunc,
              metadata: { plan_meta: planMeta },
            }),
          )
        }
      }

      trace.update({ output: { role: "assistant", content: assistantText } })
    }),
  )
}

// ── Provider lifecycle ───────────────────────────────

const createProvider = (config: Config) => {
  const processor = new LangfuseSpanProcessor({
    publicKey: config.publicKey,
    secretKey: config.secretKey,
    baseUrl: config.host,
    timeout: 10,
    exportMode: "immediate",
  })

  const provider = new NodeTracerProvider({ spanProcessors: [processor] })
  provider.register()
  setLangfuseTracerProvider(provider)

  return {
    provider,
    async [Symbol.asyncDispose]() {
      await provider.shutdown()
    },
  }
}

// ── Main ─────────────────────────────────────────────

const main = async () => {
  const config = readConfig()
  if (!config) return 0

  const payload = await readPayload()
  if (!payload) return 0

  const event = (payload.hook_event_name ?? "unknown").padEnd(
    "PostToolUseFailure".length,
  )
  using _ = timed(`${event} (session=${payload.session_id})`)
  await using __ = createProvider(config)

  await using state = await openState(payload.session_id)

  const [msgs, nextState] = await readNewMessages(
    payload.transcript_path,
    state,
  )
  Object.assign(state, nextState)

  if (!msgs.length) return 0

  const turns = assembleTurns(msgs).toArray()
  if (!turns.length) return 0

  for (const [i, turn] of turns.entries()) {
    try {
      emitTurn({
        sessionId: payload.session_id,
        turnNum: state.turnCount + i + 1,
        turn,
        transcriptPath: payload.transcript_path,
      })
    } catch (e) {
      log({ level: "error", msg: String(e) })
    }
  }

  state.turnCount += turns.length

  return 0
}

exit(await main())

File: opt/claude-code/hooks/langfuse-trace.py
#!/usr/bin/env -S -- PYTHONSAFEPATH= python3
from __future__ import annotations

from contextlib import contextmanager, nullcontext, suppress
from dataclasses import dataclass, replace
from datetime import datetime, timezone
from hashlib import sha256
from json import dumps, loads
from logging import DEBUG, INFO, basicConfig, captureWarnings, getLogger
from os import environ
from os import replace as atomic_replace
from pathlib import Path
from sys import exit, stdin
from time import time
from typing import Any, Iterator, NotRequired, TypedDict

from langfuse import Langfuse, propagate_attributes

with nullcontext():
    captureWarnings(True)
    basicConfig(
        format="%(message)s",
        level=DEBUG if environ.get("CC_LANGFUSE_DEBUG", "").lower() == "true" else INFO,
    )

_MAX_CHARS = int(environ.get("CC_LANGFUSE__MAX_CHARS", "20000"))
_ROOT = Path(__file__).resolve().parent.parent.parent.parent
_SESSIONS_DIR = _ROOT / "var" / "sessions"


@contextmanager
def _log_errors() -> Iterator[None]:
    try:
        yield
    except Exception as e:
        getLogger().error("%s", e, exc_info=True)


@contextmanager
def _timed(label: str) -> Iterator[None]:
    getLogger().debug("%s", f"{label} started")
    start = time()
    yield
    getLogger().debug("%s", f"{label} completed in {time() - start:.2f}s")


# ── Types ──────────────────────────────────────────────


class _ApiMessage(TypedDict, total=False):
    role: str
    content: Any
    model: str
    id: str


class _TranscriptLine(TypedDict, total=False):
    type: str
    message: _ApiMessage
    content: Any


class _TruncMeta(TypedDict):
    truncated: bool
    orig_len: int
    kept_len: NotRequired[int]
    sha256: NotRequired[str]


class _ToolCall(TypedDict):
    id: str
    name: str
    input: Any
    output: NotRequired[str | None]
    output_meta: NotRequired[_TruncMeta]


@dataclass(frozen=True)
class _Config:
    public_key: str
    secret_key: str
    host: str


@dataclass(frozen=True)
class _HookPayload:
    session_id: str
    transcript_path: Path


@dataclass(frozen=True)
class _ReaderState:
    offset: int = 0
    buffer: str = ""
    turn_count: int = 0


@dataclass(frozen=True)
class _Turn:
    user_msg: _TranscriptLine
    assistant_msgs: tuple[_TranscriptLine, ...]
    tool_results: dict[str, Any]


# ── Input ──────────────────────────────────────────────


def _read_config() -> _Config | None:
    if environ.get("TRACE_TO_LANGFUSE", "").lower() != "true":
        return None

    public_key = environ.get("CC_LANGFUSE_PUBLIC_KEY") or environ.get(
        "LANGFUSE_PUBLIC_KEY"
    )
    secret_key = environ.get("CC_LANGFUSE_SECRET_KEY") or environ.get(
        "LANGFUSE_SECRET_KEY"
    )
    if not public_key or not secret_key:
        return None

    host = (
        environ.get("CC_LANGFUSE_BASE_URL")
        or environ.get("LANGFUSE_BASE_URL")
        or "https://cloud.langfuse.com"
    )
    return _Config(public_key=public_key, secret_key=secret_key, host=host)


def _read_payload() -> _HookPayload | None:
    try:
        data = stdin.read()
        if not data.strip():
            return None
        payload = loads(data)
    except Exception as e:
        getLogger().error("%s", e, exc_info=True)
        return None

    session_id = (
        payload.get("sessionId")
        or payload.get("session_id")
        or payload.get("session", {}).get("id")
    )
    transcript = (
        payload.get("transcriptPath")
        or payload.get("transcript_path")
        or payload.get("transcript", {}).get("path")
    )
    if not session_id or not transcript:
        getLogger().debug("%s", "Missing session_id or transcript_path; exiting.")
        return None

    try:
        transcript_path = Path(transcript).expanduser().resolve()
    except Exception as e:
        getLogger().error("%s", e, exc_info=True)
        return None

    if not transcript_path.exists():
        getLogger().debug("%s", f"Transcript does not exist: {transcript_path}")
        return None

    return _HookPayload(session_id=session_id, transcript_path=transcript_path)


# ── State ──────────────────────────────────────────────


def _state_path(session_id: str) -> Path:
    return _SESSIONS_DIR / f"{session_id}.langfuse.json"


def _load_state(session_id: str) -> _ReaderState:
    p = _state_path(session_id)
    with _log_errors():
        if p.exists():
            raw = loads(p.read_text(encoding="utf-8"))
            return _ReaderState(
                offset=int(raw.get("offset", 0)),
                buffer=str(raw.get("buffer", "")),
                turn_count=int(raw.get("turn_count", 0)),
            )
    return _ReaderState()


@contextmanager
def _atomic_write(path: Path) -> Iterator[Path]:
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(".tmp")
    yield tmp
    atomic_replace(tmp, path)


def _save_state(session_id: str, *, state: _ReaderState) -> None:
    with _log_errors():
        with _atomic_write(_state_path(session_id)) as tmp:
            tmp.write_text(
                dumps(
                    {
                        "offset": state.offset,
                        "buffer": state.buffer,
                        "turn_count": state.turn_count,
                        "updated": datetime.now(timezone.utc).isoformat(),
                    },
                    indent=2,
                    sort_keys=True,
                ),
                encoding="utf-8",
            )


# ── Message vocabulary ─────────────────────────────────


def _get_content(msg: _TranscriptLine) -> Any:
    match msg:
        case {"message": {"content": content}}:
            return content
        case {"content": content}:
            return content
        case _:
            return None


def _get_role(msg: _TranscriptLine) -> str | None:
    match msg:
        case {"type": "user" | "assistant" as role}:
            return str(role)
        case {"message": {"role": "user" | "assistant" as role}}:
            return str(role)
        case _:
            return None


def _content_blocks(content: Any, *, block_type: str) -> list[dict[str, Any]]:
    if not isinstance(content, list):
        return []
    return [x for x in content if isinstance(x, dict) and x.get("type") == block_type]


def _is_tool_result(msg: _TranscriptLine) -> bool:
    return _get_role(msg) == "user" and bool(
        _content_blocks(_get_content(msg), block_type="tool_result")
    )


def _get_model(msg: _TranscriptLine) -> str:
    match msg:
        case {"message": {"model": str(model)}} if model:
            return model
        case _:
            return "claude"


def _get_message_id(msg: _TranscriptLine) -> str | None:
    match msg:
        case {"message": {"id": str(mid)}} if mid:
            return mid
        case _:
            return None


def _extract_text(content: Any) -> str:
    match content:
        case str(text):
            return text
        case list(blocks):
            parts: list[str] = []
            for block in blocks:
                match block:
                    case {"type": "text", "text": str(text)} if text:
                        parts.append(text)
                    case str(text) if text:
                        parts.append(text)
            return "\n".join(parts)
        case _:
            return ""


def _truncate(s: str, *, max_chars: int = _MAX_CHARS) -> tuple[str, _TruncMeta]:
    orig_len = len(s)
    if orig_len <= max_chars:
        return s, _TruncMeta(truncated=False, orig_len=orig_len)
    head = s[:max_chars]
    return head, _TruncMeta(
        truncated=True,
        orig_len=orig_len,
        kept_len=len(head),
        sha256=sha256(s.encode("utf-8")).hexdigest(),
    )


# ── Parse ──────────────────────────────────────────────


def _read_new_messages(
    path: Path, *, state: _ReaderState
) -> tuple[list[_TranscriptLine], _ReaderState]:
    """Read only new bytes since state.offset. Returns parsed lines and updated state."""
    if not path.exists():
        return [], state

    try:
        with open(path, "rb") as f:
            f.seek(state.offset)
            chunk = f.read()
            new_offset = f.tell()
    except Exception as e:
        getLogger().error("%s", e, exc_info=True)
        return [], state

    if not chunk:
        return [], state

    text = chunk.decode("utf-8", errors="replace")
    combined = state.buffer + text
    lines = combined.split("\n")

    msgs: list[_TranscriptLine] = []
    for line in lines[:-1]:
        if not (line := line.strip()):
            continue
        with suppress(Exception):
            msgs.append(loads(line))

    return msgs, replace(state, offset=new_offset, buffer=lines[-1])


# ── Assemble ───────────────────────────────────────────


def _assemble_turns(messages: list[_TranscriptLine]) -> Iterator[_Turn]:
    """
    Group transcript rows into turns: user → assistant(s) → tool_result(s).
    Dict insertion order deduplicates assistant messages by id (latest content wins).
    """
    user_msg: _TranscriptLine | None = None
    assistants: dict[str, _TranscriptLine] = {}
    tool_results: dict[str, Any] = {}

    for msg in messages:
        if _is_tool_result(msg):
            for tr in _content_blocks(_get_content(msg), block_type="tool_result"):
                if tid := tr.get("tool_use_id"):
                    tool_results[str(tid)] = tr.get("content")
            continue

        role = _get_role(msg)

        if role == "user":
            if user_msg is not None and assistants:
                yield _Turn(
                    user_msg=user_msg,
                    assistant_msgs=tuple(assistants.values()),
                    tool_results=dict(tool_results),
                )
            user_msg = msg
            assistants = {}
            tool_results = {}
            continue

        if role == "assistant" and user_msg is not None:
            mid = _get_message_id(msg) or f"noid:{len(assistants)}"
            assistants[mid] = msg
            continue

    if user_msg is not None and assistants:
        yield _Turn(
            user_msg=user_msg,
            assistant_msgs=tuple(assistants.values()),
            tool_results=dict(tool_results),
        )


# ── Emit ───────────────────────────────────────────────


def _tool_calls(
    assistant_msgs: tuple[_TranscriptLine, ...],
) -> list[_ToolCall]:
    calls: list[_ToolCall] = []
    for am in assistant_msgs:
        for tu in _content_blocks(_get_content(am), block_type="tool_use"):
            calls.append(
                _ToolCall(
                    id=str(tu.get("id") or ""),
                    name=tu.get("name") or "unknown",
                    input=inp
                    if isinstance(
                        inp := tu.get("input"), (dict, list, str, int, float, bool)
                    )
                    else {},
                )
            )
    return calls


def _emit_turn(
    langfuse: Langfuse,
    *,
    session_id: str,
    turn_num: int,
    turn: _Turn,
    transcript_path: Path,
) -> None:
    user_text, user_meta = _truncate(_extract_text(_get_content(turn.user_msg)))
    last_assistant = turn.assistant_msgs[-1] if turn.assistant_msgs else turn.user_msg
    assistant_text, assistant_meta = _truncate(
        _extract_text(_get_content(last_assistant))
    )
    model = _get_model(turn.assistant_msgs[0] if turn.assistant_msgs else last_assistant)
    calls = _tool_calls(turn.assistant_msgs)

    for c in calls:
        if (raw := turn.tool_results.get(c["id"])) is not None:
            out_str = raw if isinstance(raw, str) else dumps(raw, ensure_ascii=False)
            out_trunc, out_meta = _truncate(out_str)
            c["output"] = out_trunc
            c["output_meta"] = out_meta
        else:
            c["output"] = None

    trace_name = f"Claude Code - Turn {turn_num}"

    with (
        propagate_attributes(
            session_id=session_id,
            trace_name=trace_name,
            tags=["claude-code"],
        ),
        langfuse.start_as_current_observation(
            name=trace_name,
            input={"role": "user", "content": user_text},
            metadata={
                "source": "claude-code",
                "session_id": session_id,
                "turn_number": turn_num,
                "transcript_path": str(transcript_path),
                "user_text": user_meta,
            },
        ) as trace,
    ):
        with langfuse.start_as_current_observation(
            name="Claude Response",
            as_type="generation",
            model=model,
            input={"role": "user", "content": user_text},
            output={"role": "assistant", "content": assistant_text},
            metadata={
                "assistant_text": assistant_meta,
                "tool_count": len(calls),
            },
        ):
            ...

        for c in calls:
            in_obj: Any = c["input"]
            in_meta: _TruncMeta | None = None
            if isinstance(in_obj, str):
                in_obj, in_meta = _truncate(in_obj)

            with langfuse.start_as_current_observation(
                name=f"Tool: {c['name']}",
                as_type="tool",
                input=in_obj,
                metadata={
                    "tool_name": c["name"],
                    "tool_id": c["id"],
                    "input_meta": in_meta,
                    "output_meta": c.get("output_meta"),
                },
            ) as tool_obs:
                tool_obs.update(output=c.get("output"))

        for c in calls:
            if c["name"] == "ExitPlanMode" and (plan := c["input"]):
                plan_str = (
                    plan if isinstance(plan, str) else dumps(plan, ensure_ascii=False)
                )
                plan_trunc, plan_meta = _truncate(plan_str)
                with langfuse.start_as_current_observation(
                    name="Plan",
                    output=plan_trunc,
                    metadata={"plan_meta": plan_meta},
                ):
                    ...

        trace.update(output={"role": "assistant", "content": assistant_text})


@contextmanager
def _langfuse_client(config: _Config) -> Iterator[Langfuse | None]:
    try:
        lf = Langfuse(
            public_key=config.public_key,
            secret_key=config.secret_key,
            host=config.host,
            timeout=10,
        )
    except Exception as e:
        getLogger().error("%s", e, exc_info=True)
        yield None
        return
    try:
        yield lf
    except Exception as e:
        getLogger().error("%s", e, exc_info=True)
    finally:
        with suppress(Exception):
            lf.shutdown()


# ── Main ───────────────────────────────────────────────


def _main() -> int:
    if (config := _read_config()) is None:
        return 0

    if (payload := _read_payload()) is None:
        return 0

    with _timed(f"hook (session={payload.session_id})"), _langfuse_client(config) as langfuse:
        if langfuse is None:
            return 0

        state = _load_state(payload.session_id)
        msgs, state = _read_new_messages(payload.transcript_path, state=state)

        if not msgs:
            _save_state(payload.session_id, state=state)
            return 0

        turns = list(_assemble_turns(msgs))
        if not turns:
            _save_state(payload.session_id, state=state)
            return 0

        emitted = 0
        for turn in turns:
            emitted += 1
            with _log_errors():
                _emit_turn(
                    langfuse,
                    session_id=payload.session_id,
                    turn_num=state.turn_count + emitted,
                    turn=turn,
                    transcript_path=payload.transcript_path,
                )

        state = replace(state, turn_count=state.turn_count + emitted)
        _save_state(payload.session_id, state=state)

        getLogger().info(
            "%s",
            f"Processed {emitted} turns (session={payload.session_id})",
        )

    return 0


exit(_main())
